larpandoracontent/LArObjects/LArAdaBoostDecisionTree.cc

```diff
@@ -0,0 +1,299 @@
+        throw pandora::StatusCodeException(pandora::STATUS_CODE_NOT_FOUND);
+    }
+
+    // Read the components of this bdt container
+    pandora::TiXmlHandle localHandle(pContainerXmlNode);
+    pandora::TiXmlElement *pCurrentXmlElement = localHandle.FirstChild().Element();
+
+    while (pCurrentXmlElement)
+    {
+        if (pandora::STATUS_CODE_SUCCESS != this->ReadComponent(pCurrentXmlElement))
+        {
+            std::cout << "AdaBoostDecisionTree: Unknown component in xml file" << std::endl;
+            throw pandora::StatusCodeException(pandora::STATUS_CODE_FAILURE);
+        }
+
+        pCurrentXmlElement = pCurrentXmlElement->NextSiblingElement();
+    }
+    m_pStrongClassifier = new StrongClassifier(m_weakClassifiers);
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+pandora::StatusCode AdaBoostDecisionTree::ReadComponent(pandora::TiXmlElement *pCurrentXmlElement)
+{
+    const std::string componentName(pCurrentXmlElement->ValueStr());
+    pandora::TiXmlHandle currentHandle(pCurrentXmlElement);
+
+    if ((std::string("Name") == componentName) || (std::string("Timestamp") == componentName))
+        return pandora::STATUS_CODE_SUCCESS;
+
+    if (std::string("DecisionTree") == componentName)
+        return this->ReadDecisionTree(currentHandle);
+
+    return pandora::STATUS_CODE_INVALID_PARAMETER;
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+pandora::StatusCode AdaBoostDecisionTree::ReadDecisionTree(pandora::TiXmlHandle &currentHandle)
+{
+    IDToNodeMap idToNodeMap;
+    double boostWeight(0.);
+    int treeIndex(0);
+
+    for (pandora::TiXmlElement *pHeadTiXmlElement = currentHandle.FirstChildElement().ToElement(); pHeadTiXmlElement != NULL; pHeadTiXmlElement = pHeadTiXmlElement->NextSiblingElement())
+    {
+        if ("TreeIndex" == pHeadTiXmlElement->ValueStr())
+        {
+            PANDORA_RETURN_RESULT_IF_AND_IF(pandora::STATUS_CODE_SUCCESS, pandora::STATUS_CODE_NOT_FOUND, !=, pandora::XmlHelper::ReadValue(currentHandle,
+                "TreeIndex", treeIndex));
+        }
+        else if ("BoostWeight" == pHeadTiXmlElement->ValueStr())
+        {
+            PANDORA_RETURN_RESULT_IF_AND_IF(pandora::STATUS_CODE_SUCCESS, pandora::STATUS_CODE_NOT_FOUND, !=, pandora::XmlHelper::ReadValue(currentHandle,
+                "BoostWeight", boostWeight));
+        }
+        else if ("Node" == pHeadTiXmlElement->ValueStr())
+        {
+            pandora::TiXmlHandle nodeHandle(pHeadTiXmlElement);
+            this->ReadNode(nodeHandle, idToNodeMap);
+        }
+    }
+
+    const WeakClassifier *pWeakClassifier = new WeakClassifier(idToNodeMap, boostWeight, treeIndex);
+    m_weakClassifiers.push_back(pWeakClassifier);
```

Comment:
Do these need to be heap allocated? Are they definitely deleted at job end? What if an exception is raised during operations? Generally safer to place local instances in a local container, so that everything is tidied up when they drop off the stack. If they need to be heap based, so be it, but be super careful with the memory management throughout. Treat every function call as if it could raise an uncaught exception; are your objects tidied up?

Reply:
Moved to stack allocation.  Still need to clean data in case of exceptions being thrown.